File: 2024/19/19.go
package main

import (
	"bufio"
	"os"
	"strings"
)

func parseInput() (map[byte][]string, []string) {
	patternMap, designs := map[byte][]string{}, []string{}
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	for _, p := range strings.Split(scanner.Text(), ", ") {
		patternMap[p[0]] = append(patternMap[p[0]], p)
	}
	scanner.Scan()
	scanner.Text()
	for scanner.Scan() {
		designs = append(designs, scanner.Text())
	}
	return patternMap, designs
}

var DFS_CACHE = map[string]int{}

func dfs(design string, patternMap map[byte][]string) int {
	if len(design) == 0 {
		return 1
	}
	if DFS_CACHE[design] > 0 {
		return DFS_CACHE[design]
	}
	count := 0
	for _, p := range patternMap[design[0]] {
		if strings.HasPrefix(design, p) {
			count += dfs(design[len(p):], patternMap)
		}
	}
	DFS_CACHE[design] = count
	return count
}

func main() {
	patternMap, designs := parseInput()

	part1, part2 := 0, 0
	for _, d := range designs {
		count := dfs(d, patternMap)
		if count > 0 {
			part1++
			part2 += count
		}
	}

	println(part1)
	println(part2)
}

File: 2024/19/19.rb
#!/usr/bin/env ruby

@cache = {}
def dfs(design, pattern_hash)
  return 1 if design.empty?
  return @cache[design] if @cache.key? design
  count = 0
  pattern_hash[design[0]].each do |p|
    count += dfs(design[p.size..], pattern_hash) if design.start_with? p
  end
  @cache[design] = count
end

patterns = $stdin.readline.chomp.split(', ')
pattern_hash = Hash.new { |h, k| h[k] = [] }
patterns.each { |p| pattern_hash[p[0]] << p }
$stdin.readline # burn empty line
designs = $stdin.readlines.map(&:chomp)

possible_designs = designs.map { |d| dfs(d, pattern_hash) }

puts possible_designs.select(&:positive?).size
puts possible_designs.sum
